pkg/referrerstore/oras/oras.go

```diff
@@ -157,62 +169,108 @@ func (store *orasStore) ListReferrers(ctx context.Context, subjectReference comm
+	} else {
+		ref := fmt.Sprintf("%s@%s", subjectReference.Path, digest)
+		resolvedBlobDesc, err = repository.Blobs().Resolve(ctx, ref)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// check if blob exists in local ORAS cache
+	isCached, err := store.localCache.Exists(ctx, resolvedBlobDesc)
 	if err != nil {
 		return nil, err
 	}
 
-	return store.getRawContentFromCache(ctx, desc)
+	if !isCached {
+		// fetch blob content from remote repository
+		rc, err := repository.Fetch(ctx, resolvedBlobDesc)
+		if err != nil {
+			return nil, err
+		}
+
+		// push fetched content to local ORAS cache
+		err = store.localCache.Push(ctx, resolvedBlobDesc, rc)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return store.getRawContentFromCache(ctx, resolvedBlobDesc)
 }
 
 func (store *orasStore) GetReferenceManifest(ctx context.Context, subjectReference common.Reference, referenceDesc ocispecs.ReferenceDescriptor) (ocispecs.ReferenceManifest, error) {
-	client, err := store.createRegistryClient(ctx, subjectReference)
+	repository, err := store.createRepository(ctx, subjectReference)
 	if err != nil {
 		return ocispecs.ReferenceManifest{}, err
 	}
-
-	var result ocispecs.ReferenceManifest
-	artifactManifestFound := false
-	_, err = oras.Graph(ctx, subjectReference.Original, referenceDesc.ArtifactType, client.Resolver,
-		func(parent artifactspec.Descriptor, parentManifest artifactspec.Manifest, objects []target.Object) error {
-			if parent.Digest == referenceDesc.Digest {
-				result = ArtifactManifestToReferenceManifest(parentManifest)
-				artifactManifestFound = true
-			}
-			return nil
-		})
-
+	var manifestBytes []byte
+	// check if manifest exists in local ORAS cache
+	isCached, err := store.localCache.Exists(ctx, referenceDesc.Descriptor)
 	if err != nil {
 		return ocispecs.ReferenceManifest{}, err
 	}
 
-	if !artifactManifestFound {
-		return ocispecs.ReferenceManifest{}, fmt.Errorf("cannot find artifact manifest with digest %s", referenceDesc.Digest)
+	if !isCached {
+		// fetch manifest content from repository
+		manifestReader, err := repository.Fetch(ctx, referenceDesc.Descriptor)
+		if err != nil {
+			return ocispecs.ReferenceManifest{}, err
+		}
+
+		manifestBytes, err = io.ReadAll(manifestReader)
+		if err != nil {
+			return ocispecs.ReferenceManifest{}, err
+		}
+
+		// push fetched manifest to local ORAS cache
+		err = store.localCache.Push(ctx, referenceDesc.Descriptor, bytes.NewReader(manifestBytes))
+		if err != nil {
+			return ocispecs.ReferenceManifest{}, err
+		}
+	} else {
+		manifestBytes, err = store.getRawContentFromCache(ctx, referenceDesc.Descriptor)
+		if err != nil {
+			return ocispecs.ReferenceManifest{}, err
+		}
+	}
+
+	// marshal manifest bytes into reference manifest descriptor
+	referenceManifest := ocispecs.ReferenceManifest{}
+	if err := json.Unmarshal(manifestBytes, &referenceManifest); err != nil {
+		return ocispecs.ReferenceManifest{}, err
 	}
 
-	return result, nil
+	return referenceManifest, nil
```

Comment:
Can this deserialized manifest be cached instead of bytes and doing the unmarshall?

Reply:
I don't think so. The local cache is a local OCI ORAS store and the ORAS push method requires the reader to be provided.